### PR TITLE
feat(ai-engine): add AI-powered resume-job matcher with fallback scoring and recommendations

### LINKED ISSUE #2715: [AI Engine] Create Resume-Job Description Matcher
### 📝 Description
Create Resume-Job Description Matcher

### 📁 File Path
`backend/src/services/resumeJobMatcher.js`

### 📋 Requirements
1. Implement core logic
2. Add error handling
3. Write unit tests
4. Ensure performance
5. Update documentation

### ✅ Acceptance Criteria
- [ ] Feature implemented according to requirements
- [ ] Tests passing
- [ ] Code reviewed
- [ ] No regressions

### 🛠️ Tech Stack
- Node.js
- NLP

### 🚀 Getting Started
1. Fork and clone the repository
2. Run `npm install` in the relevant directory
3. Create/modify the file at the path above
4. Write tests if applicable
5. Submit a PR with a clear description

**📸 COMPULSORY: You MUST include a screenshot or test output in your Pull Request description, or it will NOT be merged.**

### PR DESCRIPTION
**Pull request by 4nshhh**
## Description

Implemented an AI-powered Resume ↔ Job Description Matcher service that analyzes resume relevance against job requirements and provides actionable insights.

### Features Added

* AI-powered resume-to-job matching using configured AI providers
* Match score calculation (0–100)
* Matched skills detection
* Missing skills / skill gap analysis
* Candidate strengths identification
* Personalized improvement recommendations
* Deterministic fallback matcher when AI providers are unavailable
* Robust JSON response validation and sanitization
* Error handling for invalid inputs and AI failures
* Comprehensive JSDoc documentation

###  Testing & Validation

* Added unit tests covering:

  * Successful AI responses
  * Markdown-wrapped AI responses
  * AI failure fallback behavior
  * Input validation
  * Deterministic matching logic
* Verified Gemini API integration with real-world resume/job description examples
* Tested both strong-match and poor-match scenarios

## Type of Change

* [x] New feature
* [x] Documentation update
* [x] Performance improvement

## Related Issue

Fixes #2715

## Testing

* [x] Unit tests pass
* [x] Tested Locally
* [x] New tests added

### Test Results

```bash
node --test src/services/__tests__/resumeJobMatcher.test.js
```

Result:

```text
pass 10
fail 0
```

## Screenshots (MANDATORY for UI/UX changes)

N/A (Backend Service Feature)

### Additional Validation Screenshots Attached

1. Good-match scenario validation output

Screenshot 2026-06-08 165238

2. Low-match scenario validation output

Screenshot 2026-06-08 170847

3. Unit test execution showing all tests passing

Screenshot 2026-06-08 171443

### Service
- `backend/src/services/resumeJobMatcher.js`
  - AI-powered Resume ↔ Job Description matching service
  - Deterministic fallback matching
  - Skill gap analysis and recommendations

### Tests
- `backend/src/services/__tests__/resumeJobMatcher.test.js`
  - AI response validation tests
  - Fallback behavior tests
  - Input validation tests
  - Deterministic matching tests

## Checklist

* [x] Code follows project style guidelines
* [x] Self-reviewed my code
* [x] Added comments where necessary
* [x] Updated documentation
* [x] No new warnings generated

## Technical Notes

* Implemented AI-first architecture with deterministic fallback support.
* Added response validation to prevent malformed AI output from propagating through the system.
* Designed to remain functional even when external AI providers are unavailable.
* Optimized matching logic to maintain fast execution and predictable fallback behavior.

___

## **CodeAnt-AI Description**
**Add resume-to-job matching with AI analysis and fallback scoring**

### What Changed
- Added resume-to-job matching that returns a score, shared skills, missing skills, strengths, and improvement tips
- If AI matching fails or returns unusable output, the app now falls back to a keyword-based match so users still get results
- Empty resume text or job description now triggers a clear validation error
- Added tests for successful AI results, markdown-wrapped AI responses, fallback behavior, invalid input, and recommendation output

### Impact
`✅ Fewer matching failures when AI providers are unavailable`
`✅ Clearer resume review results`
`✅ Faster candidate screening when AI output cannot be used`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **New Features**
  * Resume-to-job matching with AI-powered analysis, deterministic fallback, and normalized match scores (clamped 0–100).
  * Automatic identification of matched/missing skills, strengths, and recommendations.
  * Input validation with clear errors for empty resume or job description; robust handling of AI response formatting.

* **Tests**
  * Added comprehensive tests covering AI responses, fenced-json parsing, fallback behavior, score bounds, and edge cases.

### CODE REVIEW

File: backend/src/services/resumeJobMatcher.js [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+Rules:
+- matchScore must be between 0 and 100
+- matchedSkills should contain relevant skills present in both resume and job description
+- missingSkills should contain important skills missing from the resume
+- strengths should contain positive observations about the candidate fit
+- recommendations should contain actionable suggestions to improve alignment
+- Return JSON only
+`;
+
+    const result = await provider.generateContent(prompt);
```
[the remark this answers is not in the document]
Comment (4nshhh) [reply backend/src/services/resumeJobMatcher.js:244]: Thanks for the suggestion. The current implementation intentionally relies on provider failures to trigger the deterministic fallback. Introducing timeout-based cancellation would require additional control-flow handling and test coverage. I'd prefer to keep this PR focused on the core matcher functionality and consider timeout handling as a follow-up enhancement.
